Core/Scripts: Migrate Tanaris NPC's to SAI

mysterious-static opened the pull request:
**Changes proposed**:
- Migrates Stone Watcher of Norgannon to SAI.
- Migrates Steward of Time to SAI.

**Target branch(es)**: 335

**Issues addressed**: Updates velinath/TrinityCore#2

**Tests performed**: SQL works against DB, server starts back up after changes.

**Known issues and TODO list**:
- [ ] Needs tested ingame. I'm planning to do this tonight/tomorrow (probably tomorrow), but wanted to put it out for everyone to review because A) more eyes are better and B) I'm pretty sure it'll work.

Commit:
Core/Scripts: Migrate Tanaris NPC's to SAI

This commit migrates Stone Watcher of Norgannon and Steward of Time to
SAI. Updates velinath/TrinityCore#2.

Commit:
get sai out move to gossip_menu_option to handle; thanks @tkrokli

Commit:
ugh oversights

tkrokli commented:
Good updates, all that remains is to test if the SAI + gossip works in-game. :+1:

velinath commented:
Sorry for the wait on this. Life came up. Hoping to have some time to test this weekend so we can get this going again.

velinath commented on sql/updates/world/3.3.5/2016_06_15_99_world.sql:
```diff
@@ -0,0 +1,30 @@
+INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`tar ...
+(7918,0,0,1,62,0,100,0,1679,0,0,0,72,0,0,0,0,0,0,7,0,0,0,0,0,0,0,'Stone Watcher of Norgannon - On Gossip Option 0 Selected - Close Gossip'),
+(7918,0,1,0,61,0,100,0,0,0,0,0,15,2954,0,0,0,0,0,7,0,0,0,0,0,0,0,'Stone Watcher of Norgannon - On Gossip Option 0 Selected - Quest Credit 2954'),
+(20142,0,0,1,62,0,100,0,9978,0,0,0,85,34891,0,0,0,0,0,7,0,0,0,0,0,0,0,'Steward of Time - On Gossip Option 0 Selected - Invoker Cast \'Flight through Caverns\''),
```
gossip cannot be selected; clicking on gossip locks window

velinath commented on sql/updates/world/3.3.5/2016_06_15_99_world.sql:
```diff
@@ -0,0 +1,30 @@
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 1674;
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` IN (9977, 9978);
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
+(15,1674,0,0,0,9,0,2954,0,0,0,0,'','Stone Watcher of Norgannon - Show Gossip Option 0 if player has accepted quest 2954'),
```
gossip menu + option do not show up even if quest is accepted

velinath replied:
tried adding 
```sql
INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES (14,1674,0,0,0,9,0,2954,0,0,0,0,'',''),
```
on the offchance that the gossip menu had to be forced (although I dont see why this would happen), and no dice.

tkrokli replied:
Maybe you need to add `gossip_menu_id` = 1674 to `creature_template`.`entry` 7918 ?

velinath replied:
Probably. I'll look at it this weekend.

velinath commented on sql/updates/world/3.3.5/2016_06_15_99_world.sql:
```diff
@@ -0,0 +1,30 @@
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` IN (9977, 9978);
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`ErrorTextId`,`ScriptName`,`Comment`) VALUES
+(15,1674,0,0,0,9,0,2954,0,0,0,0,'','Stone Watcher of Norgannon - Show Gossip Option 0 if player has accepted quest 2954'),
+(14,9977,0,0,0,14,0,10279,0,0,0,0,'','Steward of Time - Show Gossip Menu 9977 if Quest 10279 is none'),
```
gossip menu does not change if quest has not been accepted

velinath commented:
seems there are some problems with this PR, I'm rebuilding right now to see if it makes any difference, but if it does not I'd appreciate a second pair of eyes. The logic seems sound, but the actual changes are - unfortunately - not working.

velinath commented:
Yeah, behavior is still broken after recompiling.

Killyana commented:
1674 is the npc_text not the gossip menu, I cannot find the gossip that use this npc_texy (1674, 1675, 16746, 1677,1678,1679)

Killyana commented:
If you have sniffs replace gossips (57000,57001,57002,57003,57004,57005,57006);
Otherwise use this: (copy the hole sql as I fixed many things)
https://gist.github.com/Killyana/93504400062a63309f368b7523822fdb

tkrokli commented:
@velinath : review `broadcast_text`, `npc_text`, `gossip_menu` and `gossip_menu_id` to find the correct links between the texts used in the dialogue. If even one table is out of sync, it could bug out.

velinath commented:
@Killyana Oh christ yep I can't believe I got these mixed up.

I'll clean it up this weekend but this is the entire reason why everything is broken on both of these NPC's.

velinath commented:
I'll check gossip menus and see if I can find the appropriate ones otherwise will use yours. Thanks Killyana, appreciate the review / second pair of eyes.

Commit:
Update Steward of Time to proper menuid

I'm holding off on swapping around the IDs for Stone Watcher in case I find something, otherwise I'll update with the rest of @Killyana's changes soon.

Commit:
update sql

Updates to @Killyana's sql as I couldn't find menu id's either

Aokromes commented on sql/updates/world/3.3.5/2016_06_15_99_world.sql:
```diff
@@ -0,0 +1,40 @@
+
+DELETE FROM `gossip_menu_option` WHERE (`menu_id` BETWEEN 57000 AND 57005) OR (`menu_id` IN (8072));
+INSERT INTO `gossip_menu_option` (`menu_id`,`id`,`option_icon`,`option_text`,`OptionBroadcastTextID`,`option_id`,`npc_option_npcflag`,`action_menu_id`,`action_poi_id`,`box_coded`,`box_money`,`box_text`,`BoxBroadcastTextID`) VALUES
+(57000,0,0,'What function do you serve?',4054,1,3,57001,0,0,0,'',0),
```
can you include VerifiedBuild column now and mark those 57000s -1?

velinath replied:
I can't find VerifiedBuild in my gossip_menu_option table, are you saying update the broadcast_text verifiedbuild value to -1, or did this get added recently and I just need to sync my branch to 3.3.5?

tkrokli replied:
Your world DB is likely not updated. That column was added to both `gossip_menu` and `gossip_menu_option` in commit https://github.com/TrinityCore/TrinityCore/commit/37b32403270e0f1156807b0ebad64f6101f68205 4 days ago, July 7 2016, 11:20 GMT+2 (05:20 EST+DST).

velinath replied:
yeah I just have to pull then, no worries. this pr was based off of something last month

tkrokli replied:
OK. travis-ci does not look into the SQL files, so manual verification is needed.

Commit:
add VerifiedBuild

Killyana commented:
This is can be pushed.
